server/src/main/java/org/opensearch/index/mapper/IpFieldMapper.java

```diff
@@ -225,42 +226,49 @@ protected Object parseSourceValue(Object value) {
         @Override
         public Query termQuery(Object value, @Nullable QueryShardContext context) {
             failIfNotIndexedAndNoDocValues();
-            Query query;
+            final PointRangeQuery pointQuery;
             if (value instanceof InetAddress) {
-                query = InetAddressPoint.newExactQuery(name(), (InetAddress) value);
+                pointQuery = (PointRangeQuery) InetAddressPoint.newExactQuery(name(), (InetAddress) value);
             } else {
                 if (value instanceof BytesRef) {
                     value = ((BytesRef) value).utf8ToString();
                 }
                 String term = value.toString();
                 if (term.contains("/")) {
                     final Tuple<InetAddress, Integer> cidr = InetAddresses.parseCidr(term);
-                    query = InetAddressPoint.newPrefixQuery(name(), cidr.v1(), cidr.v2());
+                    pointQuery = (PointRangeQuery) InetAddressPoint.newPrefixQuery(name(), cidr.v1(), cidr.v2());
                 } else {
                     InetAddress address = InetAddresses.forString(term);
-                    query = InetAddressPoint.newExactQuery(name(), address);
+                    pointQuery = (PointRangeQuery) InetAddressPoint.newExactQuery(name(), address);
                 }
             }
-            if (isSearchable() && hasDocValues()) {
-                String term = value.toString();
-                if (term.contains("/")) {
-                    final Tuple<InetAddress, Integer> cidr = InetAddresses.parseCidr(term);
-                    return InetAddressPoint.newPrefixQuery(name(), cidr.v1(), cidr.v2());
-                }
-                return new IndexOrDocValuesQuery(
-                    query,
-                    SortedSetDocValuesField.newSlowExactQuery(name(), new BytesRef(((PointRangeQuery) query).getLowerPoint()))
+            Query dvQuery = null;
+            if (hasDocValues()) {
+                dvQuery = SortedSetDocValuesField.newSlowRangeQuery(
+                    name(),
+                    new BytesRef(pointQuery.getLowerPoint()),
+                    new BytesRef(pointQuery.getUpperPoint()),
+                    true,
+                    true
                 );
             }
-            if (hasDocValues()) {
-                String term = value.toString();
-                if (term.contains("/")) {
-                    final Tuple<InetAddress, Integer> cidr = InetAddresses.parseCidr(term);
-                    return InetAddressPoint.newPrefixQuery(name(), cidr.v1(), cidr.v2());
+            return indexOrDvQuery(pointQuery, dvQuery);
+        }
+
+        private Query indexOrDvQuery(Query pointQuery, Query dvQuery) {
```

Comment:
I know this is another attempt, I really not sure it is anyhow better: 
  - attention slips to another method
  - tons of `assert`s

The logic before was convoluted with `if`s, I don't like it either, but everything was in one place, more or less straightforward to follow, may other folks have opinions.

Reply:
I agree with the overall mood, but ready to carry on since it's called ~~thrice~~ three times. 
Other options: excessively use `java.util.Objects#requireNonNull(T)`, or inline this method three times. I'm ok with any other options, let me know how to tinker it.

Reply:
It would be nice if Java had pattern-matching on tuples in `switch` statements. Then these kinds of things could essentially be written as:

```
switch (isSearchable(), hasDocValues()) {
  case (true, true) -> new IndexOrDocValuesQuery(indexedQuery(...), docValuesQuery(...));
  case (true, false) -> indexedQuery(...);
  case (false, true) -> docValuesQuery(...);
  case (false, false) -> { throw new IllegalArgumentException(...); }
}
```

Alternatively, if we were really ambitious, I guess we could define an `enum IndexAndDocValueStatus` (a terrible name), with values (`INDEXED_ONLY`, `DV_ONLY`, `BOTH`, and `NEITHER`), and replace the two `boolean` fields in `MappedFieldType`.

Reply:
Hold my beer.
```
switch((isSearchable()?2:0)|(hasDocValues()?1:0)) {
  case 1: return dvQuery;
  case 2: return pointQuery;
  case 3: return new IndexOrDocValuesQuery(pointQuery,dvQuery)
  default: throw new AssertError("unxpctd");
}
```

Reply:
What if we convert (isSearchable()?2:0)|(hasDocValues()) to enum: SEARCHABLE, DOC_VALUES, SEARCHABLE_DOC_VALUES? (plus we could wire it up with `failIfNotIndexedAndNoDocValues()`:



```
Enum e = failIfNotIndexedAndNoDocValues();
switch(e)) {
  case SEARCHABLE: return dvQuery;
  case DOC_VALUES: return pointQuery;
  case SEARCHABLE_DOC_VALUES: return new IndexOrDocValuesQuery(pointQuery,dvQuery)
}
```

Reply:
Well, honestly I'm not afraid an if-spaghetti baked into a single method called three times. We can skin it by the following:

```
BiFunction<PointRangeQuery,Query,Query> selector = failIfNotIndexedAndNoDocValues();
..
return selector.apply(pointQ,dvQ); //  <- if-spaghetti there
``` 
also we can pass `Supplier<Query>` to reduce `if(isSearchable()/hasDv())` in `..Query()` methods. 
I suppose creating suppliers instances is insignificant in comparison to query instance creation.